connectors/sources/graphql.py

```diff
@@ -0,0 +1,333 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""GraphQL source module responsible to fetch documents based on GraphQL Query."""
+import json
+from copy import deepcopy
+from functools import cached_property
+
+import aiohttp
+from aiohttp.client_exceptions import ClientResponseError
+from graphql import parse
+
+from connectors.logger import logger
+from connectors.source import BaseDataSource, ConfigurableFieldValueError
+from connectors.utils import (
+    CancellableSleeps,
+    RetryStrategy,
+    hash_id,
+    iso_utc,
+    retryable,
+)
+
+RETRIES = 3
+RETRY_INTERVAL = 2
+
+PING_QUERY = """
+{
+  __schema {
+    queryType {
+      name
+    }
+  }
+}
+"""
+
+
+class UnauthorizedException(Exception):
+    pass
+
+
+class GraphQLClient:
+    def __init__(self, configuration):
+        self._sleeps = CancellableSleeps()
+        self.configuration = configuration
+        self._logger = logger
+        self.graphql_query = self.configuration["graphql_query"]
+        self.url = self.configuration["http_endpoint"]
+        self.graphql_objects = self.configuration["object_list"]
+        self.http_method = self.configuration["http_method"]
+        self.variables = {}
+        self.headers = {}
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    @cached_property
+    def _get_session(self):
+        timeout = aiohttp.ClientTimeout(total=300)
+        if self.configuration["authentication_method"] == "bearer":
```

Comment:
I think we should store `self.configuration["authentication_method"]` as a field inside the object as it's used at multiple places. This will then also be consistent with other fields like `self.url = self.configuration["http_endpoint"]` for example.